dftimewolf/lib/processors/gcp_cloud_resource_tree.py

```diff
@@ -0,0 +1,1034 @@
+      zone: zone where the resource are located.
+      resource_type: Resource type (currently supported types: gce_instance,
+          gce_disk, gce_image, gce_machine_image, gce_instance_template,
+          gce_snapshot)
+      mode: operational mode (online or offline).
+      resource_id: Resource id.
+      resource_name: Resource name.
+    """
+    if not resource_id and not resource_name:
+      self.ModuleError("Please supply resource_id or resource_name.",
+                       critical=True)
+    self.project_id = project_id
+    self.zone = zone
+    self.resource_id = resource_id
+    self.resource_name = resource_name
+    self.resource_type = resource_type
+    if 'offline' in mode:
+      self.mode = gcp_crt_helper.OperatingMode.OFFLINE
+    elif 'online' in mode:
+      self.mode = gcp_crt_helper.OperatingMode.ONLINE
+    else:
+      self.ModuleError("Operational mode not set.", critical=True)
+
+  def Process(self) -> None:
+    """Creates the GCP Cloud Resource Tree."""
+    # Adding this check because mypy was complaining.
+    if not self.project_id or not self.zone or not self.resource_type:
+      self.ModuleError(
+          "project_id and/or zone and/or resource_type are not set.")
+      return
+
+    if self.mode == gcp_crt_helper.OperatingMode.OFFLINE:
+      self.logger.info('Starting module in offline mode.')
+      # Get the file containers created by the previous module in the recipe
+      file_containers = self.state.GetContainers(containers.File)
+
+      # Loop over the file containers and parse the content of each to fill the
+      # resources dictionary
+      for file_container in file_containers:
+        self.logger.info(
+            f'Loading file from file system container: {file_container.path}')
+        self._ParseLogMessagesFromFileContainer(file_container)
+
+    elif self.mode == gcp_crt_helper.OperatingMode.ONLINE:
+      self.logger.info('Starting module in online mode.')
+      self._GetListOfResources(self.project_id)
+      self._GetResourcesMetaDataFromLogs(self.project_id)
+
+    else:
+      self.PublishMessage(
+          'Invalid operating mode. Supported modes are "online" or "offline"')
+
+    self._BuildResourcesParentRelationships()
+
+    if self.resource_id:
+      resource = self.resources_dict.get(self.resource_id)
+    elif self.resource_name:
+      resource = self._FindResource(self.resource_name, self.resource_type,
+                                    self.zone, self.project_id)
+
+    if not resource:
+      self.logger.error('Resource not found')
+      return
+
+    # Save resource tree to temp file
+    output_file = tempfile.NamedTemporaryFile(mode='w',
+                                              delete=False,
+                                              encoding='utf-8',
+                                              suffix='.txt')
+    output_path = output_file.name
+
+    self.PublishMessage(f'Saving resource tree to {output_path}')
```

Comment:
below you check if there is parent and if this is the case you str(parent), why do you do that? as you override __str__ and call generate tree as part of it, which loops on all parents and children then add them to the graph, won't it be enough if you just call str(parent) in all cases

Reply:
No, when you run str(reource) it only gets you the tree related to this resource, so for example you won't see the sibling of the resource. I check if the resource has a parent and if so I display the tree of the parent so sibling will be visibile which i think might be helpfull for the analyst. If the reource doesn't have a parent i just display the tree for the resource.

Reply:
@sa3eed3ed please check if I should mark this as resolved.